lsp-modeline.el

```diff
@@ -60,41 +69,50 @@
 
+(defun lsp-modeline--preferred-code-action-name (actions)
+  "Return the preferred code action name from ACTIONS."
+  (or (-some->> actions
+        (-first #'lsp:code-action-is-preferred?)
+        lsp-modeline--code-action->string)
+      (->> actions
+           lsp-seq-first
+           lsp-modeline--code-action->string)))
+
 (defun lsp-modeline--code-action->string (action)
   "Convert code ACTION to friendly string."
   (->> action
        lsp:code-action-title
        (replace-regexp-in-string "[\n\t ]+" " ")))
 
+(defun lsp-modeline--build-code-actions-segments (actions)
+  "Build the code ACTIONS string from the defined segments."
+  (mapconcat
+   (lambda (segment)
+     (pcase segment
+       ('icon (lsp-modeline--code-actions-icon))
+       ('name (propertize (lsp-modeline--preferred-code-action-name actions)
+                          'face 'lsp-modeline-code-actions-face))
+       ('count (propertize (number-to-string (seq-length actions))
+                           'face 'lsp-modeline-code-actions-face))))
+   lsp-modeline-code-actions-segments " "))
+
 (defun lsp-modeline--build-code-actions-string (actions)
   "Build the string to be presented on modeline for code ACTIONS."
-  (-let* ((icon (lsp-modeline--code-actions-icon))
-          (first-action-string (propertize (or (-some->> actions
-                                                 (-first #'lsp:code-action-is-preferred?)
-                                                 lsp-modeline--code-action->string)
-                                               (->> actions
-                                                    lsp-seq-first
-                                                    lsp-modeline--code-action->string))
-                                           'face 'lsp-modeline-code-actions-face))
-          (single-action? (= (length actions) 1))
+  (-let* ((single-action? (= (length actions) 1))
           (keybinding (concat "("
                               (-some->> #'lsp-execute-code-action
                                 where-is-internal
                                 (-find (lambda (o)
                                          (not (member (aref o 0) '(menu-bar normal-state)))))
                                 key-description)
                               ")"))
-          (string (if single-action?
-                      (concat " " icon " " first-action-string " ")
-                    (concat " " icon " " first-action-string " "
-                            (propertize (concat "(" (number-to-string (1- (seq-length actions))) " more)")
-                                        'display `((height 0.9))
-                                        'face 'lsp-modeline-code-actions-face)
-                            " "))))
-    (propertize string
+          (built-string (concat " "
+                                (lsp-modeline--build-code-actions-segments actions)
+                                " ")))
```

Comment:
I always wonder why the spaces of code action mode line seems to be a little bit weird, turn out it's because of suffix white space.
![image](https://user-images.githubusercontent.com/2631472/89445237-d610b080-d78d-11ea-8a01-1f6660c81cff.png)
Does it make sense to append the prefix space only? The next mode line components should add their prefix space already (like in this case, diagnostic)

Reply:
It makes sense, each modeline thing should add only the prefix, otherwise, this will happen, thanks for noticing it!

Reply:
`set-text-properties` replaces all the properties, `add-text-properties` works